### PR TITLE
Add user avatar using directly S3

### PR DESCRIPTION
**Pull request by laura-barluzzi**
**This PR contains a migration file.**

This PR is a simplified refactored version followed by the comments in PR Add user avatar image #1643. Part of epic User Dashboard #1491.
### Proposed changes in this pull request
1. Create the `user.avatar` field using `S3FileField` and add 1 migration file.
2. Set `default='/static/img/avatar_sm.jpg'` so that there's always an image to retrieve. Here the view with default value:

3.  In `ProfileForm` inherit `user.avatar` and create `avatar = forms.CharField()` taking the django-buckets `S3FileUploadWidget` widget class.
4.  In `ProfileForm` add a `clean_avatar` where we check if the user has removed the avatar. In case avatar is empty, we set it to the default avatar image.
5.  In `ProfileForm` add `js/image_upload.js` which enables previews of both default avatar picture (example above) and uploaded picture (example below).

6. Add feature in existing test in `accounts/tests/test_forms.py`
7. Add feature in API emulating existing case with resources.

##### Some facts of this PR

1. Currently this PR just enables users to upload, preview (in `profile.html`) and store an image file inside the `S3FileField` `user.avatar`.
2. This PR no longer uses the cropit library and does not crop the image avatar. Consequently, the original size of an image uploaded remains the same.
3. This PR allows multi-sized images to be uploaded since the majority of people don't have perfectly squared images.
4. The cropped version of images inside `user.avatar` to be applied across the website can be created via `@property decorator` such as `thumbnail` which is defined in the soon-to-be-merged PR Outreachy PDF Form generation #1409. Therefore, a discussion over the compatibility of this PR with the thumbnail functionality will be required.

### When should this PR be merged
- When is reviewed

### Risks
- Mostly browser and device compatibility

### Follow-up actions
- Create new PR after PR Outreachy PDF Form generation #1409 is _also_ merged. The new PR will adapt and apply the `thumbnail property function` to display avatar images across the platform.

#### General

- **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
  - [ ] Review 1
  - [ ] Review 2
- **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
  - [ ] Review 1
  - [ ] Review 2
- **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.
  - [ ] Review 1
  - [ ] Review 2

#### Functionality

- **Are all requirements met?** Compare implemented functionality with the requirements specification.
  - [ ] Review 1
  - [ ] Review 2
- **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 
  - [ ] Review 1
  - [ ] Review 2

#### Code

- **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
  - [ ] Review 1
  - [ ] Review 2
- **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
  - [ ] Review 1
  - [ ] Review 2
- **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 
  - [ ] Review 1
  - [ ] Review 2
- **Is the code documented sufficiently?** Large and complex classes, functions or methods must be annotated with comments following our [code-style guidelines](https://devwiki.corp.cadasta.org/Contributing%20Style%20Guide#documentation-and-comments).
  - [ ] Review 1
  - [ ] Review 2
- **Has the scalability of this change been evaluated?**
  - [ ] Review 1
  - [ ] Review 2
- **Is there a maintenance plan in place?**
  - [ ] Review 1
  - [ ] Review 2

#### Tests

- **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
  - [ ] Review 1
  - [ ] Review 2
- **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
  - [ ] Review 1
  - [ ] Review 2
- **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?
  - [ ] Review 1
  - [ ] Review 2

#### Security

- **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
  - [ ] Review 1
  - [ ] Review 2
- **Are all UI and API inputs run through forms or serializers?** 
  - [ ] Review 1
  - [ ] Review 2
- **Are all external inputs validated and sanitized appropriately?**
  - [ ] Review 1
  - [ ] Review 2
- **Does all branching logic have a default case?**
  - [ ] Review 1
  - [ ] Review 2
- **Does this solution handle outliers and edge cases gracefully?**
  - [ ] Review 1
  - [ ] Review 2
- **Are all external communications secured and restricted to SSL?**
  - [ ] Review 1
  - [ ] Review 2

#### Documentation

- **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
  - [ ] Review 1
  - [ ] Review 2
- **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
  - [ ] Review 1
  - [ ] Review 2
- **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki. 
  - [ ] Review 1
  - [ ] Review 2

### CODE REVIEW

File: cadasta/accounts/tests/test_forms.py [hunk 1]
```diff
@@ -249,8 +260,11 @@ def test_update_user(self):
 
         form = forms.ProfileForm(data, request=request, instance=user)
         form.save()
-
         user.refresh_from_db()
+
+        new_file_stored_on_s3 = os.path.isfile(os.path.join(
```
Comment (oliverroick) [inline cadasta/accounts/tests/test_forms.py]: I think this isn't necessary because it's not the form's responsibility to store the file to the S3 storage.  The file is stored django-bucket's, so it's safe to assume that file is stored correctly.

File: cadasta/accounts/tests/test_serializers.py [hunk 2]
```diff
@@ -235,16 +243,20 @@ def test_create_with_valid_data(self):
...
             'measurement': 'metric',
+            'avatar': image,
         }
         serializer = serializers.UserSerializer(data=data)
         assert serializer.is_valid() is True
 
         serializer.save()
         assert User.objects.count() == 1
 
+        new_file_stored_on_s3 = os.path.isfile(os.path.join(
```
Comment (oliverroick) [inline cadasta/accounts/tests/test_serializers.py]: Same as in the form test. Not necessary to test this.

**Comment by laura-barluzzi** [comment]
Thanks @oliverroick  for the fast review. I made all the changes that you suggested and everything works just fine. I just want to clarify the reason why I ended up **not** setting the `null=True` option in the `S3FileField`.

When I tried to set `null=True` in the `S3FileField` I noticed that the conditions with `user.avatar.url` were not behaving as expected. After doing some research, I indeed read this in the [Django documentation](https://docs.djangoproject.com/en/1.11/ref/models/fields/#null):

> Avoid using null on string-based fields such as `CharField` and `TextField`. If a string-based field has `null=True`, that means it has two possible values for “no data”: `NULL`, and the empty string. In most cases, it’s redundant to have two possible values for “no data;” the Django convention is to use the empty string, not `NULL`.

Given this and that the `buckets.fields.S3FileField` is defined as a `CharField`, I had to remove the `null=True` option.

File: cadasta/accounts/models.py [hunk 3]
```diff
@@ -90,6 +94,10 @@ def get_display_name(self):
         else:
             return self.username
 
+    @property
+    def avatar_url(self):
```
Comment (oliverroick) [inline cadasta/accounts/models.py]: That's a good idea to do it this way. Can you also add tests to `test_models` to verify the behaviour of the property?  You would need to add two tests: One for an empty `avatar` field and one when it's set.

Comment (laura-barluzzi) [inline cadasta/accounts/models.py]: I added the tests you asked and from now on I'll make sure to rebase only once the PR is approved. I also added some simple CSS via the javascript file to improve a bit the UI.

**Comment by laura-barluzzi** [comment]
@oliverroick @bjohare do we need to wait for something before merging this PR?

**Comment by amplifi** [comment]
@laura-barluzzi The PR body says this should be merged after PR Outreachy PDF Form generation #1409 is merged. #1409 is pending the lat/long work, which hasn't yet been completed. Is #1409 still a dependency for this PR?

**Comment by laura-barluzzi** [comment]
@amplifi thanks for answering. This PR doesn't actively depend on the PR Outreachy PDF Form generation #1409 for being merged. This PR is independent in what it does right now.

What I meant (my mistake it came across wrongly) is that this PR still _doesn't display the stored image_ in the platform because this functionality will be built on top of the `thumbnail property` defined in the PR Outreachy PDF Form generation #1409.

In other words, this PR can be merged now. Once PR Outreachy PDF Form generation #1409 is merged, we can create a new PR that combines them so that we can add the visualization of the user avatar image across the platform (e.g. on the top nav bar or in the Members box in the dashboards of organizations and projects.

I change the description so that this is clearer.
